### PR TITLE
cilium: Send pod->node traffic through tunnel in IPSec+vxlan case

### PR DESCRIPTION
**Pull request by jrfastab**
### Packet path analysis

If a pod in host networking and/or the node itself sends traffic to a pod with encryption+vxlan enabled and encryptNode disabled we may see dropped packets. The flow is the following.

#### Forward path: pod -> node

First, a packet is sent from the host networking with srcIP=nodeIP,dstIP=podIP. Next a source IP is chosen for the packet so
that the route src will not be used. Then the route table 'routes' the packet to cilium_host using the route rule matching podIPs subnet,

    podIPSubnet via $IP dev cilium_host src $SrcIP mtu 1410

Then we drop into BPF space with srcIP=nodeIP,dstIP=podIP as above. Here we do an ipcache lookup for the destination. The ipcache lookup will have a hit for the podIP and will have both a key and tunnelIP. For example something like this,

    10.128.5.129/32     1169 3 10.0.0.4

Using above key identifier, in the example '3', the bpf_host program will mark the packet for encryption. This will pass encryption parameters through the ctx where the ingress program attached to cilium_net will in turn use those parameters to set skb->mark before passing up to the stack for encryption. At this point we have a skb ingress'ing the stack with an skb->mark=0xe00,srcIP=nodeIP,dstIP=podIP.

Here is where the trouble starts. This will miss encryption policy rules because unless encryptNode is enabled we do not have a rule to match srcIP=nodeIP. So the unencrypted packet is sent back to cilium_host using a route in the encryption routing table. Then finally bpf_host will send the skb to the overlay, cilium_vxlan.

The packet is then received by the podIP node and sent to cilium_vxlan because its a vxlan packet. The vxlan header is popped off and the inner (unencrypted) packet is sent to the pod, remember srcIP=nodeIP still.

#### Response path: node -> pod

Assuming the above was a SYN packet the pod will respond with a SYN/ACK with srcIP=podIP,dstIP=nodeIP. This will be handled by the bpf_lxc program.

First, we will do an ipcache lookup and get an ipcache entry, but this entry will not have a tunnelIP (although it will have a key). Because no tunnelIP is available a tunnel map lookup will be done. This will fail because the dstIP is a nodeIP and not in a tunnel subnet. (The tunnel map key is done by masking the dstIP with the subnet.)

Because the program did not find a valid tunnel the packet is sent to the stack. The stack will run through iptables, but because the packet flow is asymmetric (SYN over cilium_vxlan, SYN/ACK over nic) the MASQUERADE rules will be skipped. The end result is we try to send a packet with the srcIP=podIP.

Here a couple different outcomes are possible. If the network infra is strict the packet may be dropped at the sending node, refusing to send a packet with an unknown srcIP. If the receiving node has rp_filter=1 on the receiving interface, the packet will be dropped with a martian IP message in dmesg. Or if rp_filter=0 and the network knows how to route the srcIP somehow, it might work. For example some of my test systems managed to work without any issues.

### Fix with symmetric routing

In order to fix this we can ensure reply packets are symmetric to the request. To do this we inject tunnel info into nodeIP ipcache entry. Once this is done the pod replying will do the following,

Send a packet with srcIP=podIP,dstIP=nodeIP, this is picked up by the bpf_lxc program. The ipcache lookup finds a complete entry now with the tunnel info. Now instead of sending that packet to the stack it will be encapsulated in the vxlan header and sent/received on the original requesting node.

### Notes

A quick observation: some of the passes through the xfrm stack are useless here. We know the miss is going to happen and can signal this to the bpf layer by clearing the encryption key in the ipcache. We will do this as a follow up patch.

Open question, why does the non-encryption sender manage to send a packet with srcIP=cilium_host_IP instead of the node?

```release-note
Fix ipsec+vxlan bug where egressing packets would bypass masquerading on their way to remote nodes 
```

### CODE REVIEW

**Comment by jrfastab** [comment]
test-me-please

File: pkg/node/manager/manager.go
Comment (aanm) [inline pkg/node/manager/manager.go:369]: `s/sendiing/sending`

**Review by aanm — APPROVED** [review]
Thanks for the detailed explanation in the commit message

**Review by pchaigno — APPROVED** [review]

File: pkg/node/manager/manager.go
Comment (fristonio) [inline pkg/node/manager/manager.go]: `s/key==0/key=0`

File: pkg/node/manager/manager.go
Comment (fristonio) [inline pkg/node/manager/manager.go]:
```suggestion
		if !m.conf.NodeEncryptionEnabled() &&
```

File: pkg/option/config.go
Comment (fristonio) [inline pkg/option/config.go]: Is this comment correct? If I understand this right then we essentially decoupled `Encryption` from `NodeEncryption` right?

**Review by fristonio — APPROVED** [review]
Looks great. Amazing pull request description. :100: 
Minor nits, do not need a second review.

**Review by joestringer — APPROVED** [review]
> Open question, why does the non-encryption sender manage to send a packet with srcIP=cilium_host_IP instead of the node?

Seems like the answer is in the summary:

> First, a packet is sent from the host networking with srcIP=nodeIP,dstIP=podIP. Next a source IP is chosen for the packet so
> that the route src will not be used. Then the route table 'routes' the packet to cilium_host using the route rule matching podIPs subnet,
```
podIPSubnet via $IP dev cilium_host src $SrcIP mtu 1410
```

Before the initial route lookup, the source would not actually be known - the source is determined based on the interface that the traffic must egress. Here the route explicitly sets the source to $SrcIP .

Implementation LGTM, just minor tidyups that other reviewers already pointed out.

**Comment by brb** [comment]
Just noting, that we have the same asymmetry with vxlan w/o ipsec. pod -> host traffic escapes the tunnel, but it's a subject to the masquerade rules. I think it's worth fixing this case too, as vxlan users who have disabled masquerading might hit the same connectivity issues.

**Review by twpayne — APPROVED** [review]

File: pkg/node/manager/manager.go
Comment (jrfastab) [reply pkg/node/manager/manager.go]: C programmer at work here. ;) you don't use C style equals in your writing?

**Comment by jrfastab** [comment]
@brb Agree on fixing the general case as well, but because its an even bigger corner case there and this fix has to be backported to 1.9 and 1.8, I'll do that as follow PR and we can sort out how far to backport that one. Also I want to add tunnelIPs to all nodes regardless of encryption which will simplify things a bunch and save us from one more thing were ipsec vs !ipsec have different ipcache rules.

**Comment by jrfastab** [comment]
test-me-please

**Comment by pchaigno** [comment]
I suspect this patch will cause issues for connections from the hostns of one node to a remote pod via a NodePort service with `externalTrafficPolicy=Local`: `node1 -- VIP --> pod@node2`.

- node1 sends SYN to node2 via network (because VIP is node's IP).
- SYN is received on node2 and DNATed by kube-proxy to pod's IP. It is not SNATed because `externalTrafficPolicy=Local`.
- pod receives SYN from node1's IP.
- pod answers with SYN+ACK to node1's IP.
- SYN+ACK is sent through tunnel because of the present patch. It is thus not reverse DNATed and has srcIP=pod's IP.
- node1 receives SYN+ACK from pod's IP, answers with RST.

A service without externalTrafficPolicy=Local isn't affected because the SYN will be SNATed on node2, so the SYN+ACK will be sent to node2's IP and will go through the proper kube-proxy logic to be reverse DNATed and reverse DNATed.

File: pkg/option/config.go
Comment (jrfastab) [reply pkg/option/config.go]: no its not correct fixed by dropping node encryption and replacing with just "encryption"

**Comment by jrfastab** [comment]
Above seems problematic I guess we need to think up some solution.

**Comment by jrfastab** [comment]
Can we just do the DNAT on the host?

**Comment by brb** [comment]
> SYN+ACK is sent through tunnel because of the present patch. It is thus not reverse DNATed and has srcIP=pod's IP.

@pchaigno Why is it not reverse DNAT-ed? Because sending over the tunnel from a pod netns bypasses the netfilter's conntrack in the host netns?

One solution, which does not require complicating our datapath, is to suggest users to enable the host-reachable svc (socket-level lb; requires 4.19) and to make cilium-agent to provision  NodePort services when `--enable-node-port=false` (today it's not possible to set `--enable-node-port=true` with IPSec: https://github.com/cilium/cilium/pull/12985).

**Comment by pchaigno** [comment]
> I suspect this patch will cause issues for connections from the hostns of one node to a remote pod via a NodePort service with `externalTrafficPolicy=Local`: `node1 -- VIP --> pod@node2`.

I was able to validate this issue locally with a new test. I pushed the test at #14628.

> @pchaigno Why is it not reverse DNAT-ed? Because sending over the tunnel from a pod netns bypasses the netfilter's conntrack in the host netns?

Yep, that's my guess. We do a `redirect()` from `bpf_lxc` to `cilium_vxlan`.

> One solution, which does not require complicating our datapath, is to suggest users to enable the host-reachable svc (socket-level lb; requires 4.19) and to make cilium-agent to provision NodePort services when `--enable-node-port=false` (today it's not possible to set `--enable-node-port=true` with IPSec: #12985).

Yep, that looks like one option. The 4.19 requirement shouldn't be too much of an issue for now as we already require that for IPSec+vxlan (although we're hoping to remove that requirement). We however have a few recent requests to disable socket-level LB in some cases (e.g., for compatibility with Istio); if that becomes more frequent, it could be an issue.

Another option would be to fix the asymmetry on the DNAT, to ensure the response packets go through the correct reverse DNAT rules when they leave the pod. One inelegant way to implement that would be to send packets through the stack instead of using `redirect()` and use a new mark to route those packets to `cilium_vxlan`.

**Comment by brb** [comment]
> a new mark to route those packets

:scream:

**Comment by pchaigno** [comment]
retest-netnext

**Comment by pchaigno** [comment]
retest-4.9

**Comment by pchaigno** [comment]
retest-netnext

**Comment by pchaigno** [comment]
I've opened https://github.com/cilium/cilium/issues/14674 to track the remaining issue with this pull request, so that we can merge it once the CI is green again.

**Pull request merged**
